x-pack/plugins/endpoint/server/routes/whitelist.ts

```diff
@@ -0,0 +1,151 @@
+  router.get(
+    {
+      path: '/api/endpoint/whitelist',
+      validate: {},
+      options: { authRequired: true },
+    },
+    handleWhitelistGet
+  );
+
+  const INVALID_UUID = "`alert_id` Must be a UUID.'";
+  router.post(
+    {
+      path: '/api/endpoint/whitelist',
+      validate: {
+        body: schema.object({
+          comment: schema.maybe(schema.string()), // Optional comment explaining reason for whitelist
+          alert_id: schema.string({
+            minLength: 36, // https://tools.ietf.org/html/rfc4122#section-3
+            maxLength: 36,
+            validate(value) {
+              // Must be a UUID
+              if (!validateUUID(value)) {
+                return INVALID_UUID;
+              }
+            },
+          }),
+          file_path: schema.maybe(schema.string()),
+          signer: schema.maybe(schema.string()),
+          sha256: schema.maybe(schema.string()),
+          dismiss: schema.boolean({ defaultValue: false }), // Boolean determining if we dismiss all alerts that match this
+        }),
+      },
+      options: { authRequired: false }, // Change me
+    },
+    handleWhitelistPost
+  );
+}
+
+/**
+ * Handles the POST request for whitelist additions
+ */
+async function handleWhitelistPost(context, req, res) {
+  try {
+    // TODO check if one of whitelist fields exists
+    const alertId: string = req.body.alert_id;
+
+    const whitelistAttributeMap: Record<string, string> = {
+      file_path: 'malware.file.path',
+      process_signer: 'actingProcess.file.signer',
+      signer: 'malware.file.signer',
+      sha256: 'malware.file.hashes.sha256',
+    };
+
+    const newRules: WhitelistRule[] = [];
+    Object.keys(whitelistAttributeMap).forEach(k => {
+      if (req.body[k]) {
+        newRules.push({
+          eventTypes: [], // TODO grab the alert and get eventTypes from alert details API
+          whitelistRuleType: 'simple',
+          whitelistRule: {
+            type: 'equality',
+            value: req.body[k],
+            applyTo: whitelistAttributeMap[k],
+          },
+        });
+      }
+    });
+    const errors = await addWhitelistRule(context, newRules); // TODO handle
+    return res.ok({ body: newRules });
+  } catch (err) {
+    return res.internalError({ body: err });
+  }
+}
+
+/**
+ * Handles the GET request for whitelist retrieval
+ */
+async function handleWhitelistGet(context, req, res) {
+  try {
+    const whitelist: WhitelistRule[] = await getWhitelist(context);
+    return res.ok({ body: whitelist });
+  } catch (err) {
+    return res.internalError({ body: err });
+  }
+}
+
+/**
+ * Add a whitelist rule to the global whitelist
+ * @param ctx App context
+ * @param whitelistRules List of whitelist rules to apply
+ */
+async function addWhitelistRule(ctx, whitelistRules: Array<WhitelistRule>): Promise<boolean> {
+  let body = '';
+  whitelistRules.forEach(rule => {
+    body = body.concat(`{ "index" : {} }\n ${JSON.stringify(rule)}\n`);
```

Comment:
Does body need to be a string or can it be an object like here? https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/bulk_examples.html

Reply:
Not sure, couldn't get it to work without jsonifying it. But that doesn't preclude me from doing something wrong.